pkg/readiness/object_tracker.go

```diff
@@ -123,10 +128,36 @@ func (t *objectTracker) CancelExpect(o runtime.Object) {
 		return
 	}
 
-	delete(t.expect, k)
-	delete(t.seen, k)
-	delete(t.satisfied, k)
-	t.cancelled[k] = struct{}{}
+	obj := t.expect[k]
+	del := shouldDelete(&obj)
+	t.expect[k] = obj // need to reset to the map, because the map values are not pointers
+
+	if del {
+		delete(t.expect, k)
+		delete(t.seen, k)
+		delete(t.satisfied, k)
+		t.cancelled[k] = *t.newObjData()
+	}
+}
+
+// CancelExpect cancels an expectation and marks it so it
+// cannot be expected again going forward.
+func (t *objectTracker) CancelExpect(o runtime.Object) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	always := func(obj *objData) bool { return true }
+	t.cancelExpectNoLock(o, always)
+}
+
+func (t *objectTracker) TryCancelExpect(o runtime.Object) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
```

Comment:
Why not decrement retries here and if `shouldCancel` is true call `cancelExpectNoLock`?

Reply:
I didn't want to duplicate this section:

```
k, err := objKeyFromObject(o)            
if err != nil {                          
    log.Error(err, "skipping")           
    return                               
}                                        
```

Would it be better to duplicate it? I could also move this logic into `CancelExpect` so that this logic isn't called twice in any codepath

Reply:
IMO error handling code is just kind of duplicative anyway, I don't sweat it too much. IMO the added code complexity here is more costly than duplicating trivial code.

Personally, I'd have the interior function take an object key and duplicate the key generation on the caller function.

I'd also duplicate the `t.allSatisfied` check to occur before the call to `objKeyFromObject()` (the way it is in the original code), just to make sure the short-circuited state (active 99.9% of the time if the pod is long-lived) is as cheap as possible.

Reply:
The difference between copying those bits and copying the various deletes() is that the deletes are highly coordinated, where one missing delete could put the system in a bad state. In that situation, avoiding copying lowers the likelihood that a future change misses one copy, introducing a bug.

For the get-key logic and the short-circuit logic, the risk of duplication is annoying a programmer and lower performance, respectively.

Reply:
That's great feedback.  I like the part about trying to minimize complexity in the most crucial parts of the logic.  DRYing up trivial logic is less valuable, as that's easier to grok/harder to break.

Reply:
After moving all of this out of `cancelExpectNoLock()`, it made more sense to me to remove all conditional logic from that function.  All of the "should delete" business is now held in `TryCancelExpect()` where it belongs.  That way none of the complexity of that codepath is bleeding into the simpler `CancelExpect()`

Reply:
+1, awesome!